docs/configuration.md
# Configuration


## Basics

In Flask, the configuration system is built on top of the `app.config` attribute.
It's a dict-like attribute; thus you can operate it as a dict.

This `app.config` attribute will contain the following configuration variables:

- Flask's built-in configuration variables
- Flask extension's built-in configuration variables
- APIFlask's built-in configuration variables
- Your application's configuration variables

Here is a simple example for basic operates:

```python
from apiflask import APIFlask

app = APIFlask(__name__)

# set a config
app.config['DESCRIPTION'] = 'A wonderful API'

# read a config
description = app.config['DESCRIPTION']
```

Since it's dict-like object, you can also read a configuration variable with
a default value via the `app.config.get()` method:

```python
my_name = app.config.get('DESCRIPTION', 'default value')
```

Or update multiple values at once via `app.config.update()` method:

```python
app.config.update(
    'DESCRIPTION'='A wonderful API',
    'FOO'='bar',
    'ITEMS_PER_PAGE'=15
)
```

For a large application, you can store all the configuration variable in a separate
file, for example, a file called `settings.py`:

```python
MY_CONFIG = 'A wonderful API'
FOO = 'bar'
ITEMS_PER_PAGE = 15
```

Now you can set the configuration variables with the `app.config.from_pyfile()` method:

```python
from apiflask import APIFlask

app = APIFlask(__name__)

# set all configs from settings.py
app.config.from_pyfile('settings.py')
```

Read more about configuration management in
*[Configuration Handling][_config]{target:_blank}* in Flask's documentation.

[_config]: https://flask.palletsprojects.com/config/

!!! warning

    All configuration variables should be in uppercase.

!!! tip "Read a config outside of the application context"

    If you want to read a configuration variable outside of the application context,
    you will get an error like: "Working outside of the application context". This
    usually happened when you use an application factory.

    When you use the application factory, you can access the config with
    `current_app.config` in the view function:

    ```python
    from flask import current_app

    @bp.get('/foo')
    def foo():
        bar = current_app.config['BAR']
    ```

    However, when you define a data schema, there isn't an active application
    context, so you can't use `current_app`. In this situation, you can access
    the configration variables from the module you store them:

    ```python
    from apiflask import Schema

    from .settings import CATEGORIES  # import the configuration variable

    class PetIn(Schema):
        name = String(required=True, validate=Length(0, 10))
        category = String(required=True, validate=OneOf(CATEGORIES))  # use it
    ```

Below are all the built-in configuration variables in APIFlask.


## OpenAPI fields

All the configurations of OpenAPI-related fields will be used when generating the
OpenAPI spec. They will also be rendered by the API documentation.


### OPENAPI_VERSION

The version of OpenAPI Specification (`openapi.openapi`). This configuration can also
be configured from the `app.openapi_version` attribute.

- Type: `str`
- Default value: `'3.0.3'`
- Examples:

```python
app.config['OPENAPI_VERSION'] = '3.0.2'
```

Or:

```python
app.openapi_version = '3.0.2'
```

!!! warning "Version >= 0.4.0"

    This configuration variable was added in the [version 0.4.0](/changelog/#version-040).


### SERVERS

The server information of the API (`openapi.servers`), accepts multiple
server dicts. This configuration can also be configured from the `app.servers`
attribute.

[`AUTO_SERVERS`](/configuration#auto_servers) config is default to `True`, which
means the server information will be set automatically when the request context is available.

- Type: `List[Dict[str, str]]`
- Default value: `None`
- Examples:

```python
app.config['SERVERS'] = [
    {
        'name': 'Dev Server',
        'url': 'http://localhost:5000'
    },
    {
        'name': 'Production Server',
        'url': 'http://api.example.com'
    }
]
```

Or:

```python
app.servers = [
    {
        'name': 'Dev Server',
        'url': 'http://localhost:5000'
    },
    {
        'name': 'Production Server',
        'url': 'http://api.example.com'
    }
]
```


### TAGS

The tag list of the OpenAPI spec documentation (`openapi.tags`), accepts a
list of dicts. You can also pass a simple list contains the tag name string.
This configuration can also be configured from the `app.tags` attribute.

If not set, the blueprint names will use as tags, all the endpoints under the
blueprint will be marked with the blueprint tag automatically.

- Type: `Union[List[str], List[Dict[str, str]]]`
- Default value: `None`
- Examples:

Simple tag name list example:

```python
app.config['TAGS'] = ['foo', 'bar', 'baz']
```

With tag description:

```python
app.config['TAGS'] = [
    {'name': 'foo', 'description': 'The description of foo'},
    {'name': 'bar', 'description': 'The description of bar'},
    {'name': 'baz', 'description': 'The description of baz'}
]
```

Full OpenAPI tags example:

```python
app.config['TAGS'] = [
    {
        'name': 'foo',
        'description': 'tag description',
        'externalDocs': {
            'description': 'Find more info here',
            'url': 'http://example.com'
        }
    }
]
```

Use attribute `app.tags`:

```python
app.tags = ['foo', 'bar', 'baz']
```


### EXTERNAL_DOCS

The external documentation information of the API (`openapi.externalDocs`).
This configuration can also be configured from the `app.external_docs` attribute.

- Type: `Dict[str, str]`
- Default value: `None`
- Examples:

```python
app.config['EXTERNAL_DOCS'] = {
    'description': 'Find more info here',
    'url': 'http://docs.example.com'
}
```

Or:

```python
app.external_docs = {
    'description': 'Find more info here',
    'url': 'http://docs.example.com'
}
```


### INFO

The info field of the API (`openapi.info`). This configuration can also be configured
from the `app.info` attribute. The info object (openapi.info), it accepts a dict contains
following info fields: `description`, `termsOfService`, `contact`, `license`. You can use
separate configuration variables to overwrite this dict.

- Type: `Dict[str, str]`
- Default value: `None`
- Examples:

```python
app.config['INFO'] = {
    'description': '...',
    'termsOfService': 'http://example.com',
    'contact': {
        'name': 'API Support',
        'url': 'http://www.example.com/support',
        'email': 'support@example.com'
    },
    'license': {
        'name': 'Apache 2.0',
        'url': 'http://www.apache.org/licenses/LICENSE-2.0.html'
    }
}
```

Or:

```python
app.info = {
    'description': '...',
    'termsOfService': 'http://example.com',
    'contact': {
        'name': 'API Support',
        'url': 'http://www.example.com/support',
        'email': 'support@example.com'
    },
    'license': {
        'name': 'Apache 2.0',
        'url': 'http://www.apache.org/licenses/LICENSE-2.0.html'
    }
}
```


### DESCRIPTION

The description of the API (`openapi.info.description`). This configuration can also
be configured from the `app.description` attribute.

- Type: `str`
- Default value: `None`
- Examples:

```python
app.config['DESCRIPTION'] = 'Some description of my API.'
```

Or:

```python
app.description = 'Some description of my API.'
```


### TERMS_OF_SERVICE

The terms of service URL of the API (`openapi.info.termsOfService`).
This configuration can also be configured from the `app.terms_of_service` attribute.

- Type: `str`
- Default value: `None`
- Examples:

```python
app.config['TERMS_OF_SERVICE'] = 'http://example.com/terms/'
```

Or:

```python
app.terms_of_service = 'http://example.com/terms/'
```


### CONTACT

The contact information of the API (`openapi.info.contact`).
This configuration can also be configured from the `app.contact` attribute.

- Type: `Dict[str, str]`
- Default value: `None`
- Examples:

```python
app.config['CONTACT'] = {
    'name': 'API Support',
    'url': 'http://example.com',
    'email': 'support@example.com'
}
```

Or:

```python
app.contact = {
    'name': 'API Support',
    'url': 'http://example.com',
    'email': 'support@example.com'
}
```


### LICENSE

The license of the API (`openapi.info.license`).
This configuration can also be configured from the `app.license` attribute.

- Type: `Dict[str, str]`
- Default value: `None`
- Examples:

```python
app.config['LICENSE'] = {
    'name': 'Apache 2.0',
    'url': 'http://www.apache.org/licenses/LICENSE-2.0.html'
}
```

Or:

```python
app.license = {
    'name': 'Apache 2.0',
    'url': 'http://www.apache.org/licenses/LICENSE-2.0.html'
}
```


### SECURITY_SCHEMES

The custom security schemes of the API (`openapi.compomnents.securitySchemes`).
You should only this when you use a third-party authentication library and you
want to add OpenAPI spec for them. When using the built-in `HTTPBasicAuth`
and `HTTPTokenAuth`, the security schemes will be generated automatically.

This configuration can also be configured from the `app.security_schemes` attribute.

- Type: `Dict[str, t.Any]`
- Default value: `None`
- Examples:

```python
app.config['SECURITY_SCHEMES'] = {
    'ApiKeyAuth': {
      'type': 'apiKey',
      'in': 'header',
      'name': 'X-API-Key',
    }
}
```

Or:

```python
app.security_schemes = {
    'ApiKeyAuth': {
      'type': 'apiKey',
      'in': 'header',
      'name': 'X-API-Key',
    }
}
```


## OpenAPI spec

Customize the generation of the OpenAPI spec.


### SPEC_FORMAT

The format of the OpenAPI spec, accepts `'json'`, `'yaml'` or `'yml'`. This config
will be used in the following conditions:

- Serve the spec via the built-in route.
- Execute `flask spec` without passing the `--format`/`-f` option.

If you want to enable the YAML support, install APIFlask with the `yaml` extra
(it will install `PyYAML`):

```
$ pip install "apiflask[yaml]"
```

!!! warning
    The auto-detection of the format from the `APIFlask(spec_path=...)` was
    removed in favor of this config in 0.7.0.

- Type: `str`
- Default value: `'json'`
- Examples:

```python
app.config['SPEC_FORMAT'] = 'yaml'
```

!!! warning "Version >= 0.7.0"

    This configuration variable was added in the [version 0.7.0](/changelog/#version-070).


### LOCAL_SPEC_PATH

The path to the local OpenAPI spec file.

- Type: `str`
- Default value: `None`
- Examples:

```python
app.config['LOCAL_SPEC_PATH'] = 'openapi.json'
```

!!! warning

    If the path you passed is relative, do not put a leading slash in it.

!!! warning "Version >= 0.7.0"

    This configuration variable was added in the [version 0.7.0](/changelog/#version-070).


### LOCAL_SPEC_JSON_INDENT

The indentation of the local OpenAPI spec in JSON format.

- Type: `int`
- Default value: `2`
- Examples:

```python
app.config['LOCAL_SPEC_JSON_INDENT'] = 4
```

!!! warning "Version >= 0.7.0"

    This configuration variable was added in the [version 0.7.0](/changelog/#version-070).


### SYNC_LOCAL_SPEC

If `True`, the local spec will be in sync automatically, see the example usage at
[Keep the local spec in sync automatically](/openapi#keep-the-local-spec-in-sync-automatically).

- Type: `bool`
- Default value: `None`
- Examples:

```python
app.config['SYNC_LOCAL_SPEC'] = True
```

!!! warning "Version >= 0.7.0"

    This configuration variable was added in the [version 0.7.0](/changelog/#version-070).


### JSON_SPEC_MIMETYPE

The MIME type string for JSON OpenAPI spec response.

- Type: `str`
- Default value: `'application/json'`
- Examples:

```python
app.config['JSON_SPEC_MIMETYPE'] = 'application/custom-json'
```

!!! warning "Version >= 0.4.0"

    This configuration variable was added in the [version 0.4.0](/changelog/#version-040).


### YAML_SPEC_MIMETYPE

The MIME type string for YAML OpenAPI spec response.

- Type: `str`
- Default value: `'text/vnd.yaml'`
- Examples:

```python
app.config['YAML_SPEC_MIMETYPE'] = 'text/x-yaml'
```

!!! warning "Version >= 0.4.0"

    This configuration variable was added in the [version 0.4.0](/changelog/#version-040).


### SPEC_PROCESSOR_PASS_OBJECT

If `True`, the `spec` argument passed to the spec processor will be an
[`apispec.APISpec`](https://apispec.readthedocs.io/en/latest/api_core.html#apispec.APISpec) object.

- Type: `bool`
- Default value: `False`
- Examples:

```python
app.config['SPEC_PROCESSOR_PASS_OBJECT'] = True
```

!!! warning "Version >= 1.3.0"

    This configuration variable was added in the [version 1.3.0](/changelog/#version-130).


### DOCS_DECORATORS

The custom decorators of the OpenAPI documentation UI endpoint (`/docs`).

- Type: `List[Callable]`
- Default value: `None`
- Examples:

```python
app.config['DOCS_DECORATORS'] = [your_decorator]
```


### SPEC_DECORATORS

The custom decorators of the OpenAPI spec endpoint (`/openapi.json`).

- Type: `List[Callable]`
- Default value: `None`
- Examples:

```python
app.config['SPEC_DECORATORS'] = [your_decorator]
```


### SWAGGER_UI_OAUTH_REDIRECT_DECORATORS

The custom decorators of the Swagger UI OAuth redirect endpoint (`/docs/oauth2-redirect`).

- Type: `List[Callable]`
- Default value: `None`
- Examples:

```python
app.config['SWAGGER_UI_OAUTH_REDIRECT_DECORATORS'] = [your_decorator]
```


## Automation behavior control

The following configuration variables are used to control the automation behavior
of APIFlask. The default values of all these configuration variables are `True`,
you can disable them if you needed.


### AUTO_TAGS

Enable or disable auto tags (`openapi.tags`) generation from the name of the blueprint.

!!! tip

    This automation behavior only happens when `app.tags` or config `TAGS` is not set.

- Type: `bool`
- Default value: `True`
- Examples:

```python
app.config['AUTO_TAGS'] = False
```


### AUTO_SERVERS

!!! warning "Version >= 1.2.1"

    This configuration variable was added in the [version 1.2.1](/changelog/#version-121).

Enable or disable auto servers (`openapi.servers`) generation from the request context.

It's useful when you are running the API behind a reverse proxy with a URL prefix.

Notice the `servers` field will not exist when the request context is not available (e.g. `flask spec` command).

!!! tip

    This automation behavior only happens when `app.servers` or config `SERVERS` is not set.

- Type: `bool`
- Default value: `True`
- Examples:

```python
app.config['AUTO_SERVERS'] = False
```


### AUTO_OPERATION_SUMMARY

Enable or disable auto path summary from the name or docstring of the view function.

!!! tip

    This automation behavior only happens when the view function doesn't decorate
    with `@app.doc(summary=...)`.

- Type: `bool`
- Default value: `True`
- Examples:

```python
app.config['AUTO_OPERATION_SUMMARY'] = False
```

!!! warning

    This variable was renamed from `AUTO_PATH_SUMMARY` to `AUTO_OPERATION_SUMMARY`
    since version 0.8.0.


### AUTO_OPERATION_DESCRIPTION

Enable or disable auto path description from the docstring of the view function.

!!! tip

    This automation behavior only happens when the view function doesn't decorate
    with `@app.doc(description=...)`.

- Type: `bool`
- Default value: `True`
- Examples:

```python
app.config['AUTO_OPERATION_DESCRIPTION'] = False
```

!!! warning

    This variable was renamed from `AUTO_PATH_DESCRIPTION` to `AUTO_OPERATION_DESCRIPTION`
    since version 0.8.0.


### AUTO_OPERATION_ID

!!! warning "Version >= 0.10.0"

    This feature was added in the [version 0.10.0](/changelog/#version-0100).

Enable or disable auto operationId from the method and endpoint of the view function.
See more details in [Set `operationId`](/openapi/#set-operationid).

- Type: `bool`
- Default value: `False`
- Examples:

```python
app.config['AUTO_OPERATION_ID'] = True
```


### AUTO_200_RESPONSE

If a view function doesn't decorate with either `@app.input`, `@app.output`, `@app.auth_required`
or `@app.doc`, APIFlask will add a default 200 response for this view into OpenAPI spec.
Set this config to `False` to disable this behavior.

!!! tip

    You can change the description of the default 200 response with config
    `SUCCESS_DESCRIPTION`.

- Type: `bool`
- Default value: `True`
- Examples:

```python
app.config['AUTO_200_RESPONSE'] = False
```


### `AUTO_404_RESPONSE`

If a view function's URL rule contains a variable. By default, APIFlask will add a
404 response for this view into OpenAPI spec. Set this config to `False` to disable
this behavior.

!!! tip

    You can change the description of the automatic 404 response with config
    `NOT_FOUND_DESCRIPTION`.

- Type: `bool`
- Default value: `True`
- Examples:

```python
app.config['AUTO_404_RESPONSE'] = False
```

!!! warning "Version >= 0.8.0"

    This configuration variable was added in the [version 0.8.0](/changelog/#version-080).


### AUTO_VALIDATION_ERROR_RESPONSE

If a view function uses `@app.input` to validate input request data, APIFlask will add a
validation error response into OpenAPI spec for this view. Set this config to `False`
to disable this behavior.

- Type: `bool`
- Default value: `True`
- Examples:

```python
app.config['AUTO_VALIDATION_ERROR_RESPONSE'] = False
```


### AUTO_AUTH_ERROR_RESPONSE

If a view function uses `@app.auth_required` to restrict the access, APIFlask will add
an authentication error response into OpenAPI spec for this view. Set this
config to `False` to disable this behavior.

- Type: `bool`
- Default value: `True`
- Examples:

```python
app.config['AUTO_AUTH_ERROR_RESPONSE'] = False
```


## Response customization

The following configuration variables are used to customize auto-responses.


### SUCCESS_DESCRIPTION

The default OpenAPI description of the 2XX responses.

- Type: `str`
- Default value: `Successful response`
- Examples:

```python
app.config['SUCCESS_DESCRIPTION'] = 'Success!'
```

!!! warning "Version >= 0.4.0"

    This configuration variable was added in the [version 0.4.0](/changelog/#version-040).


### NOT_FOUND_DESCRIPTION

The default OpenAPI description of the 404 response.

- Type: `str`
- Default value: `Not found`
- Examples:

```python
app.config['NOT_FOUND_DESCRIPTION'] = 'Missing'
```

!!! warning "Version >= 0.8.0"

    This configuration variable was added in the [version 0.8.0](/changelog/#version-080).


### VALIDATION_ERROR_STATUS_CODE

The status code of validation error response.

- Type: `int`
- Default value: `422`
- Examples:

```python
app.config['VALIDATION_ERROR_STATUS_CODE'] = 422
```


### VALIDATION_ERROR_DESCRIPTION

The OpenAPI description of validation error response.

- Type: `str`
- Default value: `'Validation error'`
- Examples:

```python
app.config['VALIDATION_ERROR_DESCRIPTION'] = 'Invalid JSON body'
```


### VALIDATION_ERROR_SCHEMA

The schema of validation error response, accepts a schema class or
a dict of OpenAPI schema definition.

- Type: `Union[Schema, dict]`
- Default value: `apiflask.schemas.validation_error_schema`
- Examples:

```python
app.config['VALIDATION_ERROR_SCHEMA'] = CustomValidationErrorSchema
```


### AUTH_ERROR_STATUS_CODE

The status code of authentication error response.

- Type: `int`
- Default value: `401`
- Examples:

```python
app.config['AUTH_ERROR_STATUS_CODE'] = 403
```


### AUTH_ERROR_DESCRIPTION

The OpenAPI description of authentication error response.

- Type: `str`
- Default value: `'Authentication error'`
- Examples:

```python
app.config['AUTH_ERROR_DESCRIPTION'] = 'Auth error'
```


### HTTP_ERROR_SCHEMA

The schema of generic HTTP error response, accepts a schema class or
a dict of OpenAPI schema definition.

- Type: `Union[Schema, dict]`
- Default value: `apiflask.schemas.http_error_schema`
- Examples:

```python
app.config['HTTP_ERROR_SCHEMA'] = CustomHTTPErrorSchema
```


### BASE_RESPONSE_SCHEMA

The schema of base response schema, accepts a schema class or a dict of
OpenAPI schema definition.

- Type: `Union[Schema, dict]`
- Default value: `None`
- Examples:

```python
from apiflask import APIFlask, Schema
from apiflask.fields import String, Integer, Field

app = APIFlask(__name__)

class BaseResponse(Schema):
    message = String()
    status_code = Integer()
    data = Field()

app.config['BASE_RESPONSE_SCHEMA'] = BaseResponse
```

!!! warning "Version >= 0.9.0"

    This configuration variable was added in the [version 0.9.0](/changelog/#version-090).


### BASE_RESPONSE_DATA_KEY

The data key of the base response, it should match the data field name in the base
response schema.

- Type: `str`
- Default value: `'data'`
- Examples:

```python
app.config['BASE_RESPONSE_DATA_KEY'] = 'data'
```

!!! warning "Version >= 0.9.0"

    This configuration variable was added in the [version 0.9.0](/changelog/#version-090).

!!! note

    `BASE_RESPONSE_DATA_KEY` shoud be the same as `data_key` instead of the name of the property if you specify `data_key`.

    For example, `BASE_RESPONSE_DATA_KEY` shoud be assigned `payload` instead of `data` in the base response schema.
    ```python
    class BaseResponse(Schema):
        message = String()
        status_code = Integer()
        data = Field(data_key='payload')
    ```

## API documentation

The following configuration variables used to customize API documentation.


### DOCS_FAVICON

The absolute or relative URL of the favicon image file of API documentations.

- Type: `str`
- Default value: `'https://apiflask.com/_assets/favicon.png'`
- Examples:

```python
app.config['DOCS_FAVICON'] = 'https://cdn.example.com/favicon.png'
```


### REDOC_USE_GOOGLE_FONT

Enable or disable Google font in Redoc documentation.

- Type: `bool`
- Default value: `True`
- Examples:

```python
app.config['REDOC_USE_GOOGLE_FONT'] = False
```


### REDOC_CONFIG

The configuration options pass to Redoc. See the available options in the
[Redoc documentation](https://github.com/Redocly/redoc#redoc-options-object).

- Type: `dict`
- Default value: `None`
- Examples:

```python
app.config['REDOC_CONFIG'] = {'disableSearch': True, 'hideLoading': True}
```

!!! warning "Version >= 0.9.0"

    This configuration variable was added in the [version 0.9.0](/changelog/#version-090).


### REDOC_STANDALONE_JS`

The absolute or relative URL of the Redoc standalone JavaScript file.

- Type: `str`
- Default value: `'https://cdn.jsdelivr.net/npm/redoc@next/bundles/redoc.standalone.js'`
- Examples:

```python
app.config['REDOC_STANDALONE_JS'] = 'https://cdn.redoc.ly/redoc/latest/bundles/redoc.standalone.js'
```


### SWAGGER_UI_CSS

The absolute or relative URL of the Swagger UI CSS file.

- Type: `str`
- Default value: `'https://cdn.jsdelivr.net/npm/swagger-ui-dist@3/swagger-ui.css'`
- Examples:

```python
app.config['SWAGGER_UI_CSS'] = 'https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/4.11.1/swagger-ui.min.css'
```


### SWAGGER_UI_BUNDLE_JS

The absolute or relative URL of the Swagger UI bundle JavaScript file.

- Type: `str`
- Default value: `'https://cdn.jsdelivr.net/npm/swagger-ui-dist@3/swagger-ui-bundle.js'`
- Examples:

```python
app.config['SWAGGER_UI_BUNDLE_JS'] = 'https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/4.11.1/swagger-ui-bundle.min.js'
```


### SWAGGER_UI_STANDALONE_PRESET_JS

The absolute or relative URL of the Swagger UI standalone preset JavaScript file.

- Type: `str`
- Default value: `'https://cdn.jsdelivr.net/npm/swagger-ui-dist@3/swagger-ui-standalone-preset.js'`
- Examples:

```python
app.config['SWAGGER_UI_STANDALONE_PRESET_JS'] = 'https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/4.11.1/swagger-ui-standalone-preset.min.js'
```


### SWAGGER_UI_LAYOUT

The layout of Swagger UI, one of `'BaseLayout'` and `'StandaloneLayout'`.

- Type: `str`

CHANGES.md
## Version 2.3.3

Released: -

- Add 'docs_oauth2_redirect_path_external' parameter to support absolute redirect url ([issue #602][issue_602]).
- Update the redoc's default settings to use jsDelivr by default instead ([pr #650](pr_650)).

[issue_602]: https://github.com/apiflask/apiflask/issues/602
[pr_650]: https://github.com/apiflask/apiflask/pull/650


## Version 2.3.2

Released: 2024/12/15

- Fix response headers to be compliant with the OpenAPI specification for versions 3.0.0+ ([issue #631][issue_631]).
- Fix input data loading implementation when input validation is skipped ([issue #629][issue_629]).

[issue_629]: https://github.com/apiflask/apiflask/issues/629
[issue_631]: https://github.com/apiflask/apiflask/issues/631


## Version 2.3.1

Released: 2024/12/7

- Include input documentation in API spec when specifying `validation=False` on `@input` decorator  ([issue #626][issue_626]).

[issue_626]: https://github.com/apiflask/apiflask/issues/626


## Version 2.3.0

Released: 2024/11/25

- Support skipping the validation for the request body with `@input(validation=False)` ([issue #573][issue_573]).
- Enable CI test for Python 3.13.

[issue_573]: https://github.com/apiflask/apiflask/issues/573


## Version 2.2.1

Released: 2024/9/8

- Fix OpenAPI spec for multiple content types with `json_or_form` location ([issue #600][issue_600]).

[issue_600]: https://github.com/apiflask/apiflask/issues/600


## Version 2.2.0

Released: 2024/8/3

- Fix deprecated warnings in tests ([issue #594][issue_594]).
- Use postponed evaluation of annotations ([pr #585][pr_585]).
- Add static OpenAPI docs example ([issue #587][issue_587]).
- Add spec extensions support with `@doc(extensions=...)` for the view function ([issue #571][issue_571]).

[pr_585]: https://github.com/apiflask/apiflask/pull/585
[issue_594]: https://github.com/apiflask/apiflask/issues/594
[issue_587]: https://github.com/apiflask/apiflask/issues/587
[issue_571]: https://github.com/apiflask/apiflask/issues/571


## Version 2.1.3

Released: 2024/7/14

- Fix the `flask spec` command for latest Flask ([issue #582][issue_582]).

[issue_582]: https://github.com/apiflask/apiflask/issues/582


## Version 2.1.2

Released: 2024/7/6

- Deprecate the `__version__` attribute. Use feature detection or `importlib.metadata.version("apiflask")` instead.
- Explicitly skip the data serialization for response with `app.output(FileSchema)`.


## Version 2.1.1

Released: 2024/3/10

- Reuse the `File`, `Config` field, and file-related validators from flask-marshmallow ([issue #540][issue_540]).
- Add support for a `--quiet` option to the `flask spec` command ([issue #548][issue_548]).
- Fix the `flask spec` command for validators operating on complex data types ([issue #547][issue_547]).

[issue_540]: https://github.com/apiflask/apiflask/issues/540
[issue_548]: https://github.com/apiflask/apiflask/issues/548
[issue_547]: https://github.com/apiflask/apiflask/issues/547


## Version 2.1.0

Released: 2023/12/16

- Add `FileType` and `FileSize` validators ([issue #253][issue_253]).
- Allow adding multiple media types for a response ([issue #494][issue_494]).
- Support adding headers to the response schema ([issue #507][issue_507]).
- Add support for adding decorators to the OpenAPI spec and documentation UI endpoints ([issue #483][issue_483]).
- Fix the base response schema inconsistency issue.

[issue_494]: https://github.com/apiflask/apiflask/issues/494
[issue_507]: https://github.com/apiflask/apiflask/issues/507
[issue_483]: https://github.com/apiflask/apiflask/issues/483
[issue_253]: https://github.com/apiflask/apiflask/issues/253


## Version 2.0.2

Released: 2023/9/24

- Loose Flask version requirement to >= 2.0.0 and add test
  for minimum dependency versions ([issue #478][issue_478]).
- Add file uploading example application ([pr #476][pr_476]).
- Fix token authentication example application ([pr #472][pr_472]).

[issue_478]: https://github.com/apiflask/apiflask/issues/478
[pr_476]: https://github.com/apiflask/apiflask/pull/476
[pr_472]: https://github.com/apiflask/apiflask/pull/472


## Version 2.0.1

Released: 2023/8/15

- Add `Config` field to dump Flask `app.config` variables ([issue #468][issue_468]).
- Fix the typing of newly updated `doc(responses)` parameter ([issue #466][issue_466]).

[issue_466]: https://github.com/apiflask/apiflask/issues/466
[issue_468]: https://github.com/apiflask/apiflask/issues/468


## Version 2.0.0

Released: 2023/7/26<br>Codename: Gongqing

Please see the [migration guide](/migration_guide/#migrate-to-apiflask-2x) for APIFlask 1.x -> 2.0.0.

- Drop Python 3.7 support.
- Drop Flask 1.x support ([issue #442][issue_442]).
- Only pass keyword arguments to the view function. The argument name
  of the parsed data from `app.input()` will be `{location}_data` or the
  value of `arg_name` ([issue #427][issue_427]).
- Add `FileSchema` to generate OpenAPI file response ([issue #447][issue_447]).
- Support using `{}` to represent not only empty body (204) but also empty schema.
  Use `{}` or `EmptySchema` will not set the status code to 204 anymore.
- Remove the previously deprecated code:
    - The `tag` parameter in `@app.doc`.
    - The `role` parameter in `@app.auth_required`.
    - The `redoc_path` parameter in `apiflask.APIFlask` and the `/redoc` path.
- Support setting a complete response OpenAPI spec throught the `app.doc(responses)`
  parameter (i.e. `responses={400: {'description': '', 'content': ...}}`)
  ([issue #327][issue_327]).

[issue_327]: https://github.com/apiflask/apiflask/issues/327
[issue_427]: https://github.com/apiflask/apiflask/issues/427
[issue_442]: https://github.com/apiflask/apiflask/issues/442
[issue_447]: https://github.com/apiflask/apiflask/issues/447


## Version 1.3.1

Released: 2023/3/27

- Fix the import error when calling `apispec.yaml_utils.dict_to_yaml`
  ([issue #419][issue_419]).

[issue_419]: https://github.com/apiflask/apiflask/issues/419


## Version 1.3.0

Released: 2023/3/18


src/apiflask/settings.py
from __future__ import annotations

import typing as t

from .schemas import http_error_schema
from .schemas import validation_error_schema
from .types import OpenAPISchemaType
from .types import TagsType


# OpenAPI fields
OPENAPI_VERSION: str = '3.0.3'
SERVERS: list[dict[str, str]] | None = None
TAGS: TagsType | None = None
EXTERNAL_DOCS: dict[str, str] | None = None
INFO: dict[str, str | dict] | None = None
DESCRIPTION: str | None = None
TERMS_OF_SERVICE: str | None = None
CONTACT: dict[str, str] | None = None
LICENSE: dict[str, str] | None = None
SECURITY_SCHEMES: dict[str, t.Any] | None = None
# OpenAPI spec
SPEC_FORMAT: str = 'json'
YAML_SPEC_MIMETYPE: str = 'text/vnd.yaml'
JSON_SPEC_MIMETYPE: str = 'application/json'
LOCAL_SPEC_PATH: str | None = None
LOCAL_SPEC_JSON_INDENT: int = 2
SYNC_LOCAL_SPEC: bool | None = None
SPEC_PROCESSOR_PASS_OBJECT: bool = False
SPEC_DECORATORS: list[t.Callable] | None = None
DOCS_DECORATORS: list[t.Callable] | None = None
SWAGGER_UI_OAUTH_REDIRECT_DECORATORS: list[t.Callable] | None = None
# Automation behavior control
AUTO_TAGS: bool = True
AUTO_SERVERS: bool = True
AUTO_OPERATION_SUMMARY: bool = True
AUTO_OPERATION_DESCRIPTION: bool = True
AUTO_OPERATION_ID: bool = False
AUTO_200_RESPONSE: bool = True
AUTO_404_RESPONSE: bool = True
AUTO_VALIDATION_ERROR_RESPONSE: bool = True
AUTO_AUTH_ERROR_RESPONSE: bool = True
# Response customization
SUCCESS_DESCRIPTION: str = 'Successful response'
NOT_FOUND_DESCRIPTION: str = 'Not found'
VALIDATION_ERROR_DESCRIPTION: str = 'Validation error'
AUTH_ERROR_DESCRIPTION: str = 'Authentication error'
VALIDATION_ERROR_STATUS_CODE: int = 422
AUTH_ERROR_STATUS_CODE: int = 401
VALIDATION_ERROR_SCHEMA: OpenAPISchemaType = validation_error_schema
HTTP_ERROR_SCHEMA: OpenAPISchemaType = http_error_schema
BASE_RESPONSE_SCHEMA: OpenAPISchemaType | None = None
BASE_RESPONSE_DATA_KEY: str = 'data'
# API docs
DOCS_FAVICON: str = 'https://apiflask.com/_assets/favicon.png'
REDOC_USE_GOOGLE_FONT: bool = True
REDOC_STANDALONE_JS: str = 'https://cdn.redoc.ly/redoc/latest/bundles/\
redoc.standalone.js'  # TODO: rename to REDOC_JS
REDOC_CONFIG: dict | None = None
SWAGGER_UI_CSS: str = 'https://cdn.jsdelivr.net/npm/swagger-ui-dist/swagger-ui.css'
SWAGGER_UI_BUNDLE_JS: str = 'https://cdn.jsdelivr.net/npm/swagger-ui-dist/\
swagger-ui-bundle.js'  # TODO: rename to SWAGGER_UI_JS
SWAGGER_UI_STANDALONE_PRESET_JS: str = 'https://cdn.jsdelivr.net/npm/swagger-ui-dist/\
swagger-ui-standalone-preset.js'  # TODO: rename to SWAGGER_UI_STANDALONE_JS
SWAGGER_UI_LAYOUT: str = 'BaseLayout'
SWAGGER_UI_CONFIG: dict | None = None
SWAGGER_UI_OAUTH_CONFIG: dict | None = None
ELEMENTS_JS: str = 'https://cdn.jsdelivr.net/npm/@stoplight/elements/web-components.min.js'
ELEMENTS_CSS: str = 'https://cdn.jsdelivr.net/npm/@stoplight/elements/styles.min.css'
ELEMENTS_LAYOUT: str = 'sidebar'
ELEMENTS_CONFIG: dict | None = None
RAPIDOC_JS: str = 'https://cdn.jsdelivr.net/npm/rapidoc/dist/rapidoc-min.js'
RAPIDOC_THEME: str = 'light'
RAPIDOC_CONFIG: dict | None = None
RAPIPDF_JS: str = 'https://cdn.jsdelivr.net/npm/rapipdf/dist/rapipdf-min.js'
RAPIPDF_CONFIG: dict | None = None

# Version changed: 1.2.0
# Change VALIDATION_ERROR_STATUS_CODE from 400 to 422.

# Version added: 1.3.0
# SPEC_PROCESSOR_PASS_OBJECT
